src/Microsoft.Diagnostics.Monitoring.EventPipe/DiagnosticsEventPipeProcessor.cs

```diff
@@ -76,15 +78,25 @@ public async Task Process(DiagnosticsClient client, TimeSpan duration, Cancellat
                 finally
                 {
                     registration.Dispose();
+
                     EventPipeEventSource eventSource = null;
+                    EventPipeStreamProvider streamProvider = null;
+
                     lock (_lock)
                     {
                         eventSource = _eventSource;
                         _eventSource = null;
+
+                        streamProvider = _streamProvider;
+                        _streamProvider = null;
+                    }
+
+                    if (null != eventSource)
+                    {
+                        eventSource.Dispose();
                     }
 
-                    eventSource?.Dispose();
-                    if (streamProvider != null)
+                    if (null != streamProvider)
                     {
                         await streamProvider.DisposeAsync();
```

Comment:
We we not calling this before for some reason? I'm trying to figure out how the dispose didn't get called before.

Reply:
The difference that I could find that is that when invoking stopFunc and await it, it is only calling StopProcessing on the stream provider. While this starts the process of stopping, it does not wait until event processing has actually stopped. Calling dispose on the stream provider stops event processing in real time and waits for the stop to complete before returning.

There is also the possibility of calling DiagnosticsEventPipeProcessor::Process and then ::DisposeAsync without calling ::StopProcessing, which could possibly leave it in the hung state as well (since the stream provider was not disposed). However, this seemed unlikely based on how the pipeline starts and stops the processor.